### PR TITLE
supported transpose 1/2d conv and added test

### PR DESCRIPTION
**Pull request by Joustrd**
**转置卷积实现思路**：对输入和kernel重新进行整理，将其转化为普通卷积的形式；
**输入处理**：
1、根据stride在原始输入内部逐行逐列插入0；
2、转置卷积本身需要在外部加一圈kernel-1的padding;
3、对于参数padding, 转置卷积的结果需要切割掉最外围宽度为参数padding的部分，即做padding的逆操作
**kernel处理**：
1、旋转180度

**测试**：仿照卷积操作的测试标准进行了测试

### CODE REVIEW

File: paibox/synapses/base.py [hunk 1]
```diff
@@ -172,7 +178,7 @@ def __init__(
             )
 
         if order == "IOL":
-            _kernel = np.swapaxes(kernel, 0, 1)
```
Comment (KafCoppelia) [inline paibox/synapses/base.py:175]: 保留 `swapaxes`，其余地方同

File: paibox/synapses/conv_utils.py [hunk 2]
```diff
@@ -255,3 +254,204 @@ def _2d_im2col(
...
+    # we can change the transpose convolution to the form of an ordinary convolution
+
+    # Generate the transpose input array : transpose padding 0 & stride 0
+    xc_t = xc
+    xh_t = xh + (xh - 1) * (stride[0] - 1)
+    xw_t = xw + (xw - 1) * (stride[1] - 1)
+    x_transpose = np.zeros((xc_t, xh_t, xw_t), dtype=x_chw.dtype)
+    x_transpose[::1, :: stride[0], :: stride[1]] = x_chw
+    # padding 0 for transpose not for parameter padding, get new input array x_transpose
+    x_transpose = np.pad(
```
Comment (KafCoppelia) [inline paibox/synapses/conv_utils.py:431]: padding应该是 `k-p-1`

File: paibox/synapses/synapses.py [hunk 3]
```diff
@@ -138,3 +144,80 @@ def __init__(
             kernel_order,
             name=name,
         )
+
+
+class Conv1d_transpose(Conv1dSyn_transpose):
```
Comment (KafCoppelia) [inline paibox/synapses/synapses.py:149]: 转置卷积对象名称改为 `ConvTranspose1d/2d`

File: paibox/synapses/conv_utils.py [hunk 4]
```diff
@@ -255,3 +254,204 @@ def _2d_im2col(
...
+    # (ol, cin*kl) * (cin*kl, cout) = (ol, cout)
+    out = col_fm @ kernel_col.T  # + self.bias
+
+    # (ol, cout) -> (ol, cout) -> (cout, ol)
+    out = out.reshape(out_shape + (cout,)).T
+
+    return out.astype(np.int32)
+
+
+def _conv2d_transpose_faster(
```
Comment (KafCoppelia) [inline paibox/synapses/conv_utils.py:401]: 转置卷积的 `output_padding` 参数能否支持？参考：[torch.nn.ConvTranspose2d](https://pytorch.org/docs/stable/generated/torch.nn.ConvTranspose2d.html#torch.nn.ConvTranspose2d)

File: paibox/synapses/conv_utils.py [hunk 5]
```diff
@@ -255,3 +254,204 @@ def _2d_im2col(
...
+    )
+    # inverse real parameter padding
+    x_transpose = x_transpose[
+        :,
+        padding[0] : (-1 * padding[0]) if padding[0] > 0 else None,
+        padding[1] : (-1 * padding[1]) if padding[1] > 0 else None,
+    ]
+
+    # kernel: (cout, cin, kh, kw) -> (cout, cin*kh*kw)
+    kernel_rot = np.rot90(
```
Comment (KafCoppelia) [inline paibox/synapses/conv_utils.py:442]: 这里是在对kernel的 `HW` 做中心对称操作么？`np.flip` 能不能等价实现？
注释记得改

Comment (Joustrd) [reply paibox/synapses/conv_utils.py:442]: 是的，旋转180度，np.flip可以等价实现，kernel_flip = np.flip(np.flip(kernel, axis=2), axis=3)，那我将代码里的np.rot90更改为np.flip

File: paibox/synapses/conv_utils.py (hunk 2 above)
Comment (Joustrd) [reply paibox/synapses/conv_utils.py:431]: 考虑到p>k的情况(一般确实很少见)，我这里是采用先padding k-1,然后从两侧切割掉p的形式实现的

File: paibox/synapses/conv_utils.py (hunk 4 above)
Comment (Joustrd) [reply paibox/synapses/conv_utils.py:401]: 可以支持，我尽快修改

File: paibox/synapses/conv_utils.py (hunk 5 above)
Comment (KafCoppelia) [reply paibox/synapses/conv_utils.py:442]: 我觉得只需要调用一次 `np.flip` 就能完成 `np.flip(kernel, axis=(2,3)`，即只对axis=2、3做flip
参考：[numpy.flip](https://numpy.org/doc/stable/reference/generated/numpy.flip.html#numpy.flip)

**Comment by KafCoppelia** [comment]
整理一下你的提交。重新rebase bugfix-1分支，之后再应用你的修改
